presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1220,6 +1234,16 @@ protected Scope visitSetOperation(SetOperation node, Optional<Scope> scope)
             return createAndAssignScope(node, scope, outputDescriptorFields);
         }
 
+        private boolean isExpensiveDistinctUnion(SetOperation setOperation, int numOutputFields, Type outputType)
```

Comment:
You don't need to check the number of output fields every time you check for output types. So split the check into 2 parts. Check for number of fields, and check for field types. Or better, change this function to take a list of output types, and call this outside of the for loop.